torch/_refs/__init__.py

```diff
@@ -4277,6 +4278,47 @@ def diag_embed(
     return utils.mask_tensor(cond, t).contiguous()
 
 
+@register_decomposition(aten.block_diag)
+@out_wrapper()
+def _block_diag_iterable(tensors: List[TensorLikeType]) -> TensorLikeType:
+    """
+    Reference implementation of torch.block_diag
+    """
+    tensors_2d = [
+        tensor.view(1, -1) if tensor.dim() <= 1 else tensor for tensor in tensors
+    ]
+
+    ncols = builtins.sum([tensor.shape[1] for tensor in tensors_2d])
+    device = tensors_2d[0].device
+
+    result = []
+
+    col_start = 0
+    for i, tensor in enumerate(tensors_2d):
+        torch._check(
+            tensor.dim() == 2,
+            lambda: f"Input tensors must have 2 or fewer dimensions. Input {i} has {tensor.dim()} dimensions",
+        )
+        row, col = tensor.shape
+        left = torch.zeros((row, col_start), device=device, dtype=tensor.dtype)
```

Comment:
@peterbell10, the device is `meta` for these tensors and the actual device matching check happens outside of this function.

Reply:
The tensors during tracing are instances of `FakeTensor` which wraps a meta-device tensor with a "fake device" that represents the device being traced for. So when you call `tensor.device` you will see a cpu or cuda device, not meta.

Reply:
You are right. On the first pass, the devices are `meta` and then on a second pass, the actual devices are used. However in that case, the check happens before this function is called. i.e. if I add a check here and pass two tensors with different devices, I get
```
File ~/git/pytorch2/torch/_subclasses/fake_tensor.py:1271, in FakeTensor._find_common_device(func, flat_args)
   1266     raise RuntimeError(
   1267         f"Unhandled FakeTensor Device Propagation for {func}, found two different devices {common_device}, {t.device}"
   1268     )
   1270 for arg in flat_args:
-> 1271     merge_devices(arg)
   1273 # some functions that allow Python numbers to bind to Tensors
   1274 # if we have failed to find a device, and we're running one of these operators,
   1275 # we must have scalar only inputs
   1276 if should_allow_numbers_as_tensors(func) and common_device is None:
   1277     # ops with scalar only inputs always have result on cpu

File ~/git/pytorch2/torch/_subclasses/fake_tensor.py:1266, in FakeTensor._find_common_device.<locals>.merge_devices(t)
   1262     return
   1264 # mismatching devices of non-zero dim tensors, throw
   1265 # This might be valid behavior and need to be explicitly modeled, e.g. reshape_as
-> 1266 raise RuntimeError(
   1267     f"Unhandled FakeTensor Device Propagation for {func}, found two different devices {common_device}, {t.device}"
   1268 )

TorchRuntimeError: Failed running call_function <function block_diag at 0x7f9087e054c0>(*(FakeTensor(..., device='cuda:0', size=(2, 3)), FakeTensor(..., size=(5,))), **{}):
Unhandled FakeTensor Device Propagation for aten.block_diag.default, found two different devices cuda:0, cpu
```

This is true for all functions in this file. For eg: `add` doesn't have a check in `torch/_refs/__init__.py`, but if you pass two tensors with different devices, it fails early on.

Reply:
When run under dynamic shapes the error comes first from `_refs.cat`:
```python
File ~/git/pytorch/torch/_refs/__init__.py:2703, in cat(tensors, dim)
   2700 for tensor in tensors:
   2701     assert isinstance(tensor, TensorLike)
-> 2703 utils.check_same_device(*tensors, allow_cpu_scalar_tensors=False)
   2705 for t in tensors:
   2706     # match logic in legacy_cat_wrap_dim
   2707     if t.ndim == 1 and t.size(0) == 0:
```

This is a bit confusing as our code didn't call `cat` anywhere, although of course `_refs.block_diag` is higher up the traceback.  For `_refs.add` the device check lives in `_prims.add` which has the same name so there is no confusion.

This isn't the biggest deal, it just makes the error slightly less clear.

Reply:
Gotcha. Thanks for the explanation.